components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/logout/idpinit/factory/LogoutResponseFactory.java

```diff
@@ -0,0 +1,121 @@
+ */
+public class LogoutResponseFactory extends HttpIdentityResponseFactory {
+
+    private static final Log log = LogFactory.getLog(LogoutResponseFactory.class);
+
+    @Override
+    public boolean canHandle(IdentityResponse identityResponse) {
+
+        return (identityResponse instanceof LogoutResponse);
+
+    }
+
+    @Override
+    public boolean canHandle(FrameworkException exception) {
+
+        return (exception instanceof LogoutException);
+    }
+
+    @Override
+    public HttpIdentityResponse.HttpIdentityResponseBuilder create(IdentityResponse identityResponse) {
+
+        HttpIdentityResponse.HttpIdentityResponseBuilder httpIdentityResponseBuilder = new HttpIdentityResponse
+                .HttpIdentityResponseBuilder();
+        create(httpIdentityResponseBuilder, identityResponse);
+        return httpIdentityResponseBuilder;
+    }
+
+    @Override
+    public void create(HttpIdentityResponse.HttpIdentityResponseBuilder builder, IdentityResponse identityResponse) {
+
+        builder.setStatusCode(((LogoutResponse) identityResponse).getStatusCode());
+        builder.addHeader(OAuthConstants.HTTP_RESP_HEADER_CACHE_CONTROL,
+                OAuthConstants.HTTP_RESP_HEADER_VAL_CACHE_CONTROL_NO_STORE);
+        builder.addHeader(OAuthConstants.HTTP_RESP_HEADER_PRAGMA,
+                OAuthConstants.HTTP_RESP_HEADER_VAL_PRAGMA_NO_CACHE);
+        builder.addHeader(HttpHeaders.CONTENT_TYPE, MediaType.TEXT_PLAIN);
+        builder.setBody(((LogoutResponse) identityResponse).getMessage());
+    }
+
+    @Override
+    public HttpIdentityResponse.HttpIdentityResponseBuilder handleException(FrameworkException frameworkException) {
+
+        HttpIdentityResponse.HttpIdentityResponseBuilder builder;
+        if (frameworkException instanceof LogoutClientException) {
+            if (log.isDebugEnabled()) {
+                log.debug(LOGOUT_CLIENT_EXCEPTION.getMessage() + ":" + frameworkException.getMessage(),
+                        frameworkException);
+            }
+            builder = buildResponse(frameworkException.getMessage(), HttpServletResponse.SC_BAD_REQUEST);
+        } else {
+            log.error(LOGOUT_SERVER_EXCEPTION.getMessage(), frameworkException);
+            builder = buildResponse(LOGOUT_SERVER_EXCEPTION.getMessage(), HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+        }
+        return builder;
+    }
+
+    /**
+     * Build Identity response based on the error message and code.
+     *
+     * @param errorMessage
+     * @param errorCode
+     * @return
+     */
+    private HttpIdentityResponse.HttpIdentityResponseBuilder buildResponse(String errorMessage, int errorCode) {
+
+        HttpIdentityResponse.HttpIdentityResponseBuilder builder =
+                new HttpIdentityResponse.HttpIdentityResponseBuilder();
+        JSONObject responseBody = new JSONObject();
+        responseBody.appendField("Error Message", errorMessage);
+        responseBody.appendField("Error Code", errorCode);
```

Comment:
Can we make this consistent with our REST API error response?

usually we do not have spaces in json keys

Reply:
Also try to include the correlation id in the response

Reply:
A sample error response
```
{
  "code": "AAA-00000",
  "message": "Some error message.",
  "description": "Some error description.",
  "traceId": "e0fbcfeb-3617-43c4-8dd0-7b7d38e13047"
}
```